### PR TITLE
Multi AZ storage for OSD

### PR DESCRIPTION
**Pull request by anjalitelang**
This PR is for merging multi AZ storage documentation for OpenShift dedicated.

### CODE REVIEW

**Comment by anjalitelang** [comment]
@ahardin-rh  I created this PR. I dont see the preview build. Is this something you will help us with?

**Comment by anjalitelang** [comment]
@wgordon17 @spurtell Please review

File: architecture/additional_concepts/storage.adoc [hunk 1]
```diff
@@ -492,6 +492,176 @@ spec:
         claimName: myclaim
 ----
 
+ifdef::openshift-dedicated[]
+[[osd-multi-az-pv]]
+== OSD Multi-AZ Cluster PV Usage Examples
```
Comment (spurtell) [inline architecture/additional_concepts/storage.adoc]:
```suggestion
== Multi-AZ persistent volume claims
```

File: architecture/additional_concepts/storage.adoc [hunk 2]
```diff
@@ -492,6 +492,176 @@ spec:
...
+
+A walk through of common use cases for persistent volumes on a Multi-AZ
```
Comment (spurtell) [inline architecture/additional_concepts/storage.adoc]:
```suggestion
A walkthrough of common use cases for persistent volumes on a Multi-AZ
```

File: architecture/additional_concepts/storage.adoc [hunk 3]
```diff
@@ -492,6 +492,176 @@ spec:
...
+OpenShift Dedicated cluster.
+[[osd-multi-az-pv-config]]
+=== Configuration Details
```
Comment (spurtell) [inline architecture/additional_concepts/storage.adoc]:
```suggestion
=== Configuration details
```

File: architecture/additional_concepts/storage.adoc [hunk 4]
```diff
@@ -492,6 +492,176 @@ spec:
...
+From here, you can either upload your template file or copy/paste it into the
+provided text field and then click create. From here you can opt to process
+and/or save your template file to the selected namespace for future use.
+If you process the template, you’ll be taken to a page to fill out the available
+ template parameters, including the desired storage class. After populating the parameters,
+you can create the mongodb service and confirm that your PVC has been created in
+ the correct AZ.
+
+[[osd-multi-az-pvc-best-practices]]
+=== Best Practices
```
Comment (spurtell) [inline architecture/additional_concepts/storage.adoc]:
```suggestion
=== Best practices
```

**Comment by ahardin-rh** [comment]
@anjalitelang Thank you! Is this for the 3.11 OSD docs or for the 4.1 OSD docs? If it's for 4.1, we'll need to submit it against the enterprise-4.1 branch.

Assuming you have [asciibinder](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/tools_and_setup.adoc#install-asciibinder-and-dependencies) set up, you can run `asciibinder build` to see a preview of the docs.

**Comment by wgordon17** [comment]
@ahardin-rh This is for 3.11 (they shouldn't be necessary in 4 🤞 )
I'm also planning on pushing another commit to this branch to clean up the formatting a bit

**Comment by anjalitelang** [comment]
@wgordon17 thanks!

**Comment by wgordon17** [comment]
@anjalitelang can you provide me with edit access to your repo? https://github.com/anjalitelang/openshift-docs/settings/collaboration

**Comment by wgordon17** [comment]
Current preview:
screencapture-file-Users-wgordon-Projects-openshift-docs-preview-openshift-dedicated-multiazstorage-architecture-additional-concepts-storage-html-2019-05-17-12_18_18

**Comment by wgordon17** [comment]
@ahardin-rh Would you mind providing a final review?

File: architecture/additional_concepts/storage.adoc [hunk 5]
```diff
@@ -492,6 +492,183 @@ spec:
...
+[[osd-multi-az-pv]]
+== Multi-AZ persistent volume claims
+
+A walkthrough of common use cases for persistent volumes on a Multi-AZ
+{product-title} cluster.
+
+[[osd-multi-az-pv-config]]
+=== Configuration details
+
+A multi-AZ {product-title} cluster will come configured with one storage class per
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/will come/comes
s/will be set/is set

File: architecture/additional_concepts/storage.adoc [hunk 6]
```diff
@@ -492,6 +492,183 @@ spec:
...
+  accessModes:
+  - ReadWriteOnce
+  resources:
+  requests:
+    storage: 1Gi
+  storageClassName: gp2-encrypted-us-east-2b <1>
+----
+<1> Specify your storage class to override the default.
+
+Passing this manifest to `oc create -f <manifest filename>` will create a PV in
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/will create/creates

File: architecture/additional_concepts/storage.adoc [hunk 7]
```diff
@@ -492,6 +492,183 @@ spec:
...
+the AZ of the storage class provided. You can confirm this by checking the
+output of `oc get pvc`.
+
+[[osd-multi-az-pv-usecases-pvc-storageclass-ui]]
+==== Creating a PersistentVolumeClaim with a specific storage class (UI)
+
+The web console UI makes choosing a storage class for your PVC easy.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: Remove the first sentence, as we should not indicate what is easy for the user. Consider replacing with:

    To choose a storage class for your PVC using the web console, navigate to the *Storage* dashboard and click 
*Create Storage*.

File: architecture/additional_concepts/storage.adoc [hunk 8]
```diff
@@ -492,6 +492,183 @@ spec:
...
+In the application console, navigate to the Storage dashboard and click the
+"Create Storage" button. The first option on the "Create Storage" page is a
+dropdown list of available storage classes to select from.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: drop-down list of available storage classes.

File: architecture/additional_concepts/storage.adoc [hunk 9]
```diff
@@ -492,6 +492,183 @@ spec:
...
+
+image::storage-class.png[Create storage with UI,title="Storage Class dropdown"]
+
+The dropdown menu shows the name, zone, and access mode for the available
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: drop-down menu

File: architecture/additional_concepts/storage.adoc [hunk 10]
```diff
@@ -492,6 +492,183 @@ spec:
...
+storage classes. You can select your desired storage class, provide a name and
+size for the PVC, and click "Create".
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: *Create*.

File: architecture/additional_concepts/storage.adoc [hunk 11]
```diff
@@ -492,6 +492,183 @@ spec:
...
+
+[[osd-multi-az-pv-usecases-pvc-non-default-storageclass-cli]]
+==== Using non-default storage classes with OpenShift templates (CLI only)
+
+OpenShift ships with a number of templates that can be used to quickly deploy
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/OpenShift/{product-title}

File: architecture/additional_concepts/storage.adoc [hunk 12]
```diff
@@ -492,6 +492,183 @@ spec:
...
+popular applications and frameworks. These are listed in the catalog when using
+the web console and are viewable from CLI by viewing the output of.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: output of:

File: architecture/additional_concepts/storage.adoc [hunk 13]
```diff
@@ -492,6 +492,183 @@ spec:
...
+
+----
+$ oc get templates -n openshift
+----
+
+Some of these templates include PVC definitions to provide persistent storage,
+but they do not natively support selection of a non-default storage class
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: class. However, there are plans to support this in the future.

File: architecture/additional_concepts/storage.adoc [hunk 14]
```diff
@@ -492,6 +492,183 @@ spec:
...
+(however there is an RFE filed to support this in the future).
+To make use of a template with a specific storage class, you'll need to copy the
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: To use a template with a specific storage class, you need

File: architecture/additional_concepts/storage.adoc [hunk 15]
```diff
@@ -492,6 +492,183 @@ spec:
...
+template to disk and edit it prior to using it. For example, if you wanted to
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/wanted/want

File: architecture/additional_concepts/storage.adoc [hunk 16]
```diff
@@ -492,6 +492,183 @@ spec:
...
+use the MongoDB template with a PVC in us-east-2b, you would first copy the
+template to a file.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: us-east-2b: 
    
    .Copy the template to a file:
    +
    ----
    $ oc get template -n openshift mongodb-persistent -o yaml --export > mongodb-persistent.yml
    ----
   
    .  Look through . . .

File: architecture/additional_concepts/storage.adoc [hunk 17]
```diff
@@ -492,6 +492,183 @@ spec:
...
+Now that you have an updated template, you can use it from the CLI or web console.
+From the CLI, you can pass in the required parameters to `oc process` and
+create them.
+
+----
+$ oc process -f mongodb-export.yml \
+STORAGE_CLASS_NAME=gp2-encrypted-us-east-2b | oc create -f
+----
+
+Or you can upload your template via the web console and fill out the parameters
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Or/Alternatively,

File: architecture/additional_concepts/storage.adoc [hunk 18]
```diff
@@ -492,6 +492,183 @@ spec:
...
+of the template interactively. To do that, click the "Add to Project" dropdown
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: To do so, select *Import YAML / JSON* from the *Add to Project* drop-down menu in the application console.

File: architecture/additional_concepts/storage.adoc [hunk 19]
```diff
@@ -492,6 +492,183 @@ spec:
...
+in the application console and select the option "Import YAML / JSON".
+
+image::add-to-project.png[Import yaml/json to project,title="Import yaml/json to project"]
+
+From here, you can either upload your template file or copy/paste it into the
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/From here/Now,

Comment (ahardin-rh) [reply architecture/additional_concepts/storage.adoc]: copy and paste it

File: architecture/additional_concepts/storage.adoc [hunk 20]
```diff
@@ -492,6 +492,183 @@ spec:
...
+provided text field and then click create. From here you can opt to process
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: click *Create*

Comment (ahardin-rh) [reply architecture/additional_concepts/storage.adoc]: s/From here/Now,
s/you'll be taken/you are directed to a page

File: architecture/additional_concepts/storage.adoc [hunk 21]
```diff
@@ -492,6 +492,183 @@ spec:
...
+If you process the template, you'll be taken to a page to fill out the available
+template parameters, including the desired storage class. After populating the parameters,
+you can create the MongoDB service and confirm that your PVC has been created in
+the correct AZ.
+
+[[osd-multi-az-pvc-best-practices]]
+=== Best practices
+
+It is very important to deliberately select a storage class when provisioning
+PVCs in a multi-AZ OSD cluster. If every PVC is always provisioned by the
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/OSD/{product-title}

File: architecture/additional_concepts/storage.adoc [hunk 22]
```diff
@@ -492,6 +492,183 @@ spec:
...
+default storage class, this can lead to problems in the future.
+
+One of these problems is an uneven distribution of workload. If you are
+primarily running workloads that make use of persistent storage, and you always
+provision that storage using the default storage class, all of your pods will be
+scheduled to nodes in that AZ. This can lead to a situation where the nodes in one
+AZ are running out of capacity while the rest of the cluster is underutilized.
+Once you are in this state, there isn't an easy method to redistribute the workloads.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/isn't/is not

File: architecture/additional_concepts/storage.adoc [hunk 23]
```diff
@@ -492,6 +492,183 @@ spec:
...
+Any redistribution will require backing up and restoring the contents of PVs manually.
+The other problem with winding up in this state is that it effectively undermines
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/with winding up in this state/with this state
s/goes down/fails
s/go down/be lost

**Comment by ahardin-rh** [comment]
@anjalitelang @wgordon17 I made some comments. Once we get these addressed and the commits squashed, we should be good to go!

**Pull request merged**